validator_client/src/beacon_node_fallback.rs

```diff
@@ -0,0 +1,382 @@
+    /// does not remember the online state) + checks if the the node uses the correct spec (gets
+    /// remembered) + increase metric counts for each used beacon node.
+    pub async fn first_success<'a, F, O, R, Err>(
+        &'a self,
+        func: F,
+    ) -> Result<O, FallbackError<BeaconNodeError<Err>>>
+    where
+        Err: Debug,
+        F: Fn(&'a BeaconNodeHttpClient) -> R,
+        R: Future<Output = Result<O, BeaconNodeError<Err>>>,
+    {
+        let func = &func;
+        self.fallback
+            .first_success(|(node, state)| async move {
+                check_preconditions(state, |g| async move {
+                    let result = self.context.check(node).await;
+                    save_result(g, &result);
+                    report_result(node, result)
+                })
+                .await?;
+                trace!(self.context.log, "Send request"; "endpoint" => format!("{}", node));
+                report_result(node, self.context.log_error(node, func(node).await))
+            })
+            .await
+    }
+
+    /// Wrapper function for `Fallback::first_success_concurrent_retry` that checks if a beacon node
+    /// is online (but does not remember the online state) + checks if the the node uses the correct
+    /// spec (gets remembered) + increase metric counts for each used beacon node.
+    pub async fn first_success_concurrent_retry<'a, F, G, O, R, Err: Debug>(
+        &'a self,
+        func: F,
+        should_retry: G,
+        retry_delay: Duration,
+    ) -> Result<O, FallbackError<(usize, BeaconNodeError<Err>)>>
+    where
+        F: Fn(&'a BeaconNodeHttpClient) -> R,
+        R: Future<Output = Result<O, BeaconNodeError<Err>>>,
+        G: Fn(&Result<O, BeaconNodeError<Err>>) -> bool,
+    {
+        let func = &func;
+        self.fallback
+            .first_success_concurrent_retry(
+                |(node, state)| async move {
+                    check_preconditions(state, |g| async move {
+                        let result = self.context.check(node).await;
+                        save_result(g, &result);
+                        report_result(node, result)
+                    })
+                    .await?;
+                    report_result(node, self.context.log_error(node, func(node).await))
+                },
+                should_retry,
+                retry_delay,
+            )
+            .await
+    }
+}
+
+fn report_result<T, E>(endpoint: &BeaconNodeHttpClient, result: Result<T, E>) -> Result<T, E> {
+    inc_counter_vec(&ENDPOINT_REQUESTS, &[endpoint.as_ref()]);
+    if result.is_err() {
+        inc_counter_vec(&ENDPOINT_ERRORS, &[endpoint.as_ref()]);
+    }
+    result
+}
+
+fn save_result(
+    mut g: RwLockWriteGuard<Option<Result<(), BeaconNodeConnectionError>>>,
+    result: &Result<(), BeaconNodeConnectionError>,
+) {
+    if !matches!(result, Err(BeaconNodeConnectionError::Offline)) {
```

Comment:
What's the reasoning for not storing an offline check?

I might be missing something, but I thought that since we reset this state each time we check duties then it might be nice to just assume an node is offline for that period. It might protect us against delays due to timeout errors.

Reply:
Currently only the sync state gets reset regularly. Basically, I never store offline because of the discussion we had in issue #1883 where the argument was that we want the fallback to behave really as fallback without too much memory (although we needed the memory anyway, but only as short term memory). Thinking now about it, maybe it makes sense to remember and regularly reset the offline state. I will prepare a change for that.

Reply:
Ok, I restructured now the used error types to make the intentions more clear. Basically we have two sorts of memories: The `unrecoverable` and the `recoverable`. At the moment we only have one unrecoverable error and that is if the chain specs are wrong. The recoverable errors are offline and unsynced. We remember an unrecoverable error forever and only check it one time (the node must be online to check it). For recoverable errors we reset and therefore recheck them regularly.

But in the restructured code there is still a line similar to the one you commented about. This is because when checking for unrecoverable errors we may get an offline error returned which means we couldn't check the unrecoverable error. We want to propagate that offline error (even store it in the recoverable memory) but we don't want to store it in the unrecoverable memory. The new error types make this more explicit but at the cost that we need to wrap the unrecoverable error in some other error that can also represent the offline state. I hope my changes make this now clearer. If you have improvement proposals or any question I'm all ears :).